XLS-0045-prepublish-validator-lists/README.md

```diff
@@ -0,0 +1,824 @@
+`BlobInfo`, then populating the `blobs` array in the
+`TMValidatorListCollection`. Currently, the `Peer` object keeps track of
+the latest sequence sent to that peer. When building the message for the peer,
+continue to only send VLs the peer has not seen. For optimization, this may
+require grouping peers by most recent sequence before building the message(s).
+
+Before sending, check the message size, splitting it into multiple
+messages if necessary. Use `Message.getBuffer().size()`.  If the size is
+larger than the message size limit (64Mb), split the array of
+`TMValidatorListV2`s in half until it's small enough. If it gets down to
+a single `BlobInfo`, send a `TMValidatorList`. If that is still larger
+than the limit, don't send it at all.
+
+Fortunately, because the messages are distinct, no conditional
+processing needs to be done for the different messages. Processing of
+the old `TMValidatorList` does not need to change at all (other than to
+update for the changed API). The new `TMValidatorListCollection` should
+be able to build the `BlobInfo` array directly from the `blobs` member
+of the message and pass that to the updated
+`ValidatorList::applyListAndBroadcast`.
+
+#### Expiration and automatic rotation
+
+There are two functions where the VL `expiration` is currently checked:
+* `ValidatorList::expires`
+* `ValidatorList::updateTrusted`
+
+##### `ValidatorList::expires`
+
+For now, no changes will be made to `expires` to rotate VLs. When
+computing the return value, starting with `current` and continuing into
+`remaining`, find the last VL where the `effective` date overlaps with
+the `expiration` date of the previous. Use that VL's `expiration` date
+in the return value calculation.
+
+##### `ValidatorList::updateTrusted`
+
+`updateTrusted` will be modified to take the `closeTime` of the
+validated ledger (this might available in the caller
+`NetworkOPsImp::beginConsensus` as `closingInfo.parentCloseTime`, but
+check the implementation to confirm, and get the correct value if not)
+and a `const&` to the `NetworkOPsImp` calling object.
+
+For all time comparisons, if the wall clock time is more than 30 seconds
+ahead of the `closeTime`, it will be used instead.
+
+* While iterating over the publisher lists, if the first entry in `remaining`
+  has an `effective` date before the `closeTime`, then it will be moved to
+  `current`, and removed from `remaining`.
+* Additionally, the `expiration` will be compared to the `closeTime`. If
+  the `current` VL has expired and no replacement has been pulled out of
+  `remaining`, then in addition to calling `removePublisherList`, call
+  [`NetworkOPs::setUNLBlocked`](#stopping-and-resuming-operations-on-expiration).
+  (Note that `removePublisherList` is also called from
+  `ValidatorList::verify`, but that removes the publisher entirely. As
+  currently implemented, if there is another publisher configured, the
+  node will not get stuck. Is this behavior correct?)
```

Comment:
Should a rippled github issue be opened for tracking this question?

Reply:
Couldn't hurt

Reply:
@intelliot can you open this issue?